File: bcs-services/cluster-resources/pkg/handler/config/cm.go
/*
 * Tencent is pleased to support the open source community by making Blueking Container Service available.
 * Copyright (C) 2022 THL A29 Limited, a Tencent company. All rights reserved.
 * Licensed under the MIT License (the "License"); you may not use this file except
 * in compliance with the License. You may obtain a copy of the License at
 *
 * 	http://opensource.org/licenses/MIT
 *
 * Unless required by applicable law or agreed to in writing, software distributed under,
 * the License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
 * either express or implied. See the License for the specific language governing permissions and
 * limitations under the License.
 */

// Package config 配置类接口实现
package config

import (
	"context"

	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"

	"github.com/Tencent/bk-bcs/bcs-services/cluster-resources/pkg/action/perm"
	resAction "github.com/Tencent/bk-bcs/bcs-services/cluster-resources/pkg/action/resource"
	respUtil "github.com/Tencent/bk-bcs/bcs-services/cluster-resources/pkg/action/resp"
	"github.com/Tencent/bk-bcs/bcs-services/cluster-resources/pkg/action/web"
	resCsts "github.com/Tencent/bk-bcs/bcs-services/cluster-resources/pkg/resource/constants"
	"github.com/Tencent/bk-bcs/bcs-services/cluster-resources/pkg/util/pbstruct"
	clusterRes "github.com/Tencent/bk-bcs/bcs-services/cluster-resources/proto/cluster-resources"
)

// Handler ...
type Handler struct{}

// New ...
func New() *Handler {
	return &Handler{}
}

// ListCM 获取 ConfigMap 列表
func (h *Handler) ListCM(
	ctx context.Context, req *clusterRes.ResListReq, resp *clusterRes.CommonResp,
) (err error) {
	if err = perm.CheckNSAccess(ctx, req.ClusterID, req.Namespace); err != nil {
		return err
	}
	respData, err := respUtil.BuildListAPIRespData(
		ctx, respUtil.ListParams{
			req.ClusterID, resCsts.CM, req.ApiVersion, req.Namespace, req.Format, req.Scene,
		}, metav1.ListOptions{LabelSelector: req.LabelSelector},
	)
	if err != nil {
		return err
	}
	if resp.Data, err = pbstruct.Map2pbStruct(respData); err != nil {
		return err
	}
	resp.WebAnnotations, err = web.GenListConfigWebAnnos(ctx, respData)
	return err
}

// GetCM 获取单个 ConfigMap
func (h *Handler) GetCM(
	ctx context.Context, req *clusterRes.ResGetReq, resp *clusterRes.CommonResp,
) (err error) {
	if err = perm.CheckNSAccess(ctx, req.ClusterID, req.Namespace); err != nil {
		return err
	}
	respData, err := respUtil.BuildRetrieveAPIRespData(
		ctx, respUtil.GetParams{
			req.ClusterID, resCsts.CM, req.ApiVersion, req.Namespace, req.Name, req.Format,
		}, metav1.GetOptions{},
	)
	if err != nil {
		return err
	}
	if resp.Data, err = pbstruct.Map2pbStruct(respData); err != nil {
		return err
	}
	resp.WebAnnotations, err = web.GenRetrieveConfigWebAnnos(ctx, respData)
	return err
}

// CreateCM 创建 ConfigMap
func (h *Handler) CreateCM(
	ctx context.Context, req *clusterRes.ResCreateReq, resp *clusterRes.CommonResp,
) (err error) {
	resp.Data, err = resAction.NewResMgr(req.ClusterID, "", resCsts.CM).Create(
		ctx, req.RawData, req.Format, true, metav1.CreateOptions{},
	)
	return err
}

// UpdateCM 更新 ConfigMap
func (h *Handler) UpdateCM(
	ctx context.Context, req *clusterRes.ResUpdateReq, resp *clusterRes.CommonResp,
) (err error) {
	resp.Data, err = resAction.NewResMgr(req.ClusterID, "", resCsts.CM).Update(
		ctx, req.Namespace, req.Name, req.RawData, req.Format, metav1.UpdateOptions{},
	)
	return err
}

// DeleteCM 删除 ConfigMap
func (h *Handler) DeleteCM(
	ctx context.Context, req *clusterRes.ResDeleteReq, _ *clusterRes.CommonResp,
) error {
	return resAction.NewResMgr(req.ClusterID, "", resCsts.CM).Delete(
		ctx, req.Namespace, req.Name, metav1.DeleteOptions{},
	)
}

File: bcs-ui/backend/tests/testing_utils/mocks/cm.py
# -*- coding: utf-8 -*-
#
# Tencent is pleased to support the open source community by making 蓝鲸智云PaaS平台社区版 (BlueKing PaaS Community Edition) available.
# Copyright (C) 2017-2019 THL A29 Limited, a Tencent company. All rights reserved.
# Licensed under the MIT License (the "License"); you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://opensource.org/licenses/MIT
#
# Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
# an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
# specific language governing permissions and limitations under the License.
#
from typing import Dict, List

from django.utils.crypto import get_random_string


class StubClusterManagerClient:
    def __init__(self, *args, **kwargs):
        pass

    def list_cloud_accounts(self, project_id: str) -> List[Dict]:
        """查询云账号"""
        return [
            {
                "projectID": project_id,
                "accountID": f"BCS-tencentCloud-{get_random_string(8)}",
                "accountName": "operator测试凭证",
                "account": {"secretID": get_random_string(8), "secretKey": get_random_string(8)},
                "creator": get_random_string(8),
                "updater": get_random_string(8),
                "creatTime": "2022-06-11T13:05:29+08:00",
                "updateTime": "2022-06-11T13:05:29+08:00",
            },
            {
                "projectID": project_id,
                "accountID": f"BCS-tencentCloud-{get_random_string(8)}",
                "accountName": "ingress-controller测试凭证",
                "account": {"secretID": get_random_string(8), "secretKey": get_random_string(8)},
                "creator": get_random_string(8),
                "updater": get_random_string(8),
                "creatTime": "2022-06-12T13:05:29+08:00",
                "updateTime": "2022-06-12T13:05:29+08:00",
            },
        ]

    def list_cloud_accounts_by_ids(self, accounts_ids: List[str]) -> List[Dict]:
        """根据 ID 查询账号信息"""
        return [
            {
                "projectID": get_random_string(32),
                "accountID": acct_id,
                "accountName": "operator测试凭证",
                "account": {"secretID": get_random_string(8), "secretKey": get_random_string(8)},
                "creator": get_random_string(8),
                "updater": get_random_string(8),
                "creatTime": "2022-06-11T13:05:29+08:00",
                "updateTime": "2022-06-11T13:05:29+08:00",
            }
            for acct_id in accounts_ids
        ]
